parser.c

```diff
@@ -642,63 +575,101 @@ void process_pihole_log(int file)
 			// Free allocated memory
 			free(client);
 			free(domain);
-			free(domainwithspaces);
 		}
-		else if(strstr(readbuffer,": forwarded") != NULL)
+		// is this a "gravity.list" line?
+		else if(strstr(readbuffer,"/gravity.list ") != NULL && strstr(readbuffer," is ") != NULL)
 		{
-			// Check if this domain names contains only printable characters
-			// if not: skip analysis of this log line
-			if(strstr(readbuffer,"<name unprintable>") != NULL)
+			// Check query for invalid characters
+			if(!checkQuery(readbuffer, "gravity.list"))
+				continue;
+
+			// Get dnsmasq's ID for this transaction
+			int dnsmasqID = getID(readbuffer);
+			// Skip invalid lines
+			if(dnsmasqID < 0)
+				continue;
+
+			// Save forwardID in corresponding query indentified by dnsmasq's ID
+			bool found = false;
+			for(i=0; i<counters.queries; i++)
 			{
-				if(debug) logg("Ignoring <name unprintable> domain (forwarded)");
+				// Check both UUID and generation of this query
+				if(queries[i].id == dnsmasqID && queries[i].generation == loggeneration)
+				{
+					queries[i].status = 1;
+					found = true;
+					break;
+				}
+			}
+			if(!found)
+			{
+				// This may happen e.g. if the original query was a PTR query or "pi.hole"
+				// as we ignore them altogether
 				continue;
 			}
 
-			// Check if this is a PTR query
-			// if so: skip analysis of this log line
-			if(strstr(readbuffer,"in-addr.arpa") != NULL)
+			if(!queries[i].complete)
+			{
+				// This query is no longer unknown
+				counters.unknown--;
+				// ... but got blocked
+				counters.blocked++;
+				// Hereby, this query is now fully determined
+				queries[i].complete = true;
+
+				// Get time index
+				int timeidx = getTimeIndex(readbuffer);
+				validate_access("overTime", timeidx, true, __LINE__, __FUNCTION__, __FILE__);
+				overTime[timeidx].blocked++;
+				validate_access("domains", queries[i].domainID, true, __LINE__, __FUNCTION__, __FILE__);
+				domains[queries[i].domainID].blockedcount++;
+			}
+		}
+		// is this a "forwarded" line?
+		else if(strstr(readbuffer," forwarded ") != NULL && strstr(readbuffer," to ") != NULL)
+		{
+			// Check query for invalid characters
+			if(!checkQuery(readbuffer, "forwarded"))
+				continue;
+
+			// Get dnsmasq's ID for this transaction
+			int dnsmasqID = getID(readbuffer);
+			// Skip invalid lines
+			if(dnsmasqID < 0)
 				continue;
 
 			// Get ID of forward destination, create new forward destination record
 			// if not found in current data structure
 			int forwardID = getforwardID(readbuffer, true);
 			if(forwardID == -2)
+			{
+				if(debug) logg("Skipping malformated forwarded line");
 				continue;
+			}
 
-			// Get timestamp
-			int querytimestamp, overTimetimestamp, timeidx = -1, i;
-			extracttimestamp(readbuffer, &querytimestamp, &overTimetimestamp);
-
+			// Save forwardID in corresponding query indentified by dnsmasq's ID
```

Comment:
I think this is where that comment came from (see above).

Reply:
I agree :-)